File: problem17/solution.go
package main

import (
    "fmt"
    "strconv"
)


var units = map[int]string{
    1: "one", 2: "two", 3: "three", 4: "four", 5: "five", 
    6: "six", 7:"seven", 8:"eight", 9: "nine", 10: "ten", 
    11: "eleven", 12: "twelve", 13: "thirteen", 14: "fourteen",
    15: "fifteen", 16: "sixteen", 17: "seventeen", 18: "eighteen",
    19: "nineteen",
}

var tens = map[int]string{
    2: "twenty", 3: "thirty", 4: "forty", 5: "fifty", 
    6: "sixty", 7:"seventy", 8:"eighty", 9: "ninety",
}


func print_number(number int) {
    if number < 20 {
        fmt.Println(units[number])
    } else {
        var number_str = strconv.Itoa(number)
        var first_number,_ = strconv.Atoi(number_str[0:1])
        var remainder_number, _ = strconv.Atoi(number_str[1:len(number_str)])
        var first_word = ""
        if (len(number_str) == 3){
            first_word = units[first_number] + " hundred"
        }
        if (len(number_str) == 2){
            first_word = tens[first_number]

        }
        fmt.Print(first_word,  " ")
        print_number(remainder_number)


        //for _, r := range number_str {
        //    fmt.Println(string(r))
       // }
    }
}

func main() {
    for i := 1; i < 400; i++ {
        print_number(i)
    }
}
    
    

File: problem17/solution.py
########################################################################################################################
# Problem 17: Lexicographic permutations
########################################################################################################################

# If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19
# letters used in total.

# If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


# NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115
# (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with
# British usage
########################################################################################################################
# NOTES
# 
#
########################################################################################################################

from common.util import LOG

below_twenty = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", 
                "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
tens = ["ignore", "ignore", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]


def get_string_repr(i):
    i_str = str(i)
    if i < 20:
        str_repr = below_twenty[i]
    elif len(i_str) == 2:
        str_repr = tens[int(i_str[0])] + below_twenty[int(i_str[1])]
    elif len(i_str) == 3:
        str_repr = below_twenty[int(i_str[0])] + "hundred"
        if i % 100 > 0:
            str_repr += "and" + get_string_repr(int(i_str[1:]))
    elif i == 1000:
        str_repr = "onethousand"
    return str_repr

def run(args):
    letter_count = 0
    for i in range(1, 1001):
        str_repr = get_string_repr(i)
        LOG.debug(f"{i} {str_repr} {len(str_repr)}")
        letter_count += len(str_repr)

    print("SOLUTION:", letter_count)

